sympy/simplify/cse_diff.py

```diff
@@ -0,0 +1,188 @@
+
+        See Also
+        ========
+
+        Direct Acyclic Graph : https://en.wikipedia.org/wiki/Directed_acyclic_graph
+
+    """
+
+    # Check Inputs
+    if not isinstance(wrt, (MatrixBase, list, tuple)):
+        raise TypeError("``wrt`` must be an iterable of variables")
+
+    elif isinstance(wrt, (list, tuple)):
+        wrt = Matrix(wrt)
+
+    if not isinstance(expr, MatrixBase):
+        raise TypeError("``expr`` must be of matrix type")
+
+    # Both ``wrt`` and ``expr`` can be a row or a column matrix, so we need to make
+    # sure all valid combinations work, but everything else fails:
+    if not (expr.shape[0] == 1 or expr.shape[1] == 1):
+        raise TypeError("``expr`` must be a row or a column matrix")
+
+    if not (wrt.shape[0] == 1 or wrt.shape[1] == 1):
+        raise TypeError("``wrt`` must be a row or a column matrix")
+
+    replacements, reduced_expr = cse(expr)
+    replacements, reduced_expr = postprocess(replacements, reduced_expr)
+
+    if replacements:
+        rep_sym, sub_expr = map(Matrix, zip(*replacements))
+    else:
+        rep_sym, sub_expr, reduced_expr = Matrix([]), Matrix([]), [expr]
+
+    l_sub, l_wrt, l_red = len(sub_expr), len(wrt), len(reduced_expr[0])
+
+    f1 = {
+        (i, j): diff_value
+        for i, r in enumerate(reduced_expr[0])
+        for j, w in enumerate(wrt)
+        if (diff_value := r.diff(w)) != 0
+    }
+
+    if not replacements:
+        return SparseMatrix(l_red, l_wrt, f1)
+
+    f2 = {
+        (i, j): diff_value
+        for i, (r, fs) in enumerate([(r, r.free_symbols) for r in reduced_expr[0]])
+        for j, s in enumerate(rep_sym)
+        if s in fs and (diff_value := r.diff(s)) != 0
+    }
+
+    symbols = expr.free_symbols
+    precomputed_fs = [s.free_symbols - symbols for s in sub_expr]
+
+    C = Counter({(0, j): diff_value for j, w in enumerate(wrt) if (diff_value := sub_expr[0].diff(w)) != 0})
+
+    for i in range(1, l_sub):
+        Bi = {(i, j): diff_value for j in range(i + 1)
+              if rep_sym[j] in precomputed_fs[i] and (diff_value := sub_expr[i].diff(rep_sym[j])) != 0}
+
+        Ai = Counter({(i, j): diff_value for j, w in enumerate(wrt)
+                      if (diff_value := sub_expr[i].diff(w)) != 0})
+
+        if Bi:
+            Ci = dok_matrix_multiply(Bi, C)
+            Ci.update(Ai)  # Use Counter's update method to add Ai to Ci
+            C.update(Ci)  # Update C with the result
+        else:
+            C.update(Ai)
+
+    J = dok_matrix_multiply(f2, C)
+
+    for (i, j), value in f1.items():
+        J[(i, j)] += value
+
+    sub_rep = dict(replacements)
+    for i, ik in enumerate(precomputed_fs):
+        sub_dict = {j: sub_rep[j] for j in ik}
+        sub_rep[rep_sym[i]] = sub_rep[rep_sym[i]].xreplace(sub_dict)
+
+    J = {key: expr.xreplace(sub_rep) for key, expr in J.items()}
+    J = SparseMatrix(None, J)
```

Comment:
Seems a bit odd to return a `SparseMatrix`. Shouldn't it return a normal SymPy Matrix?

Reply:
Should be the same type as the input.

Reply:
A normal SymPy matrix is internally sparse and already has sparse matrix multiplication. The same is true fo SparseMatrix which has the same implementation internally as Matrix. There is never any reason to use SparseMatrix now but it is kept for compatibility.

I suspect that the reason for not using it here is because there is this which is a bit slower than normal multiplication:
https://github.com/sympy/sympy/blob/e30c119507fa64d4d8f7e78c09bf1649abb2e9b3/sympy/polys/matrices/sdm.py#L1453-L1471
https://github.com/sympy/sympy/blob/e30c119507fa64d4d8f7e78c09bf1649abb2e9b3/sympy/polys/matrices/sdm.py#L1497-L1503
This is what is used for normal matrix multiplication even though `sdm_matmul` would otherwise be faster. The faster routine assumes that `0*expr` evaluates to `0` for all expressions which is not true for all expressions in SymPy. The faster routine is used for domains like `QQ[x,y]` but not when the domain is EXRAW as is the default for matrices that are not purely rational numbers.

Rather than using other types here or reimplementing what `DomainMatrix` already does it would be better just to add a method to Matrix to compute a matrix product without trying to handle things like `0*oo`. The implementation is already there but it is not used by default to preserve existing behaviour like:
```python
In [1]: Matrix([[x, oo]]) * Matrix([0, 0])
Out[1]: [nan]
```
Also though if the expressions in the matrix are at all large than I doubt that all the zero multiplies are going to make a lot of difference to the total time for the matrix product. Possibly multiplication by zero can be optimised as well.

Reply:
> The implementation is already there but it is not used by default to preserve existing behaviour like

Could you point me where this method is implemented? 

@oscarbenjamin could you also explain me more practically what objects and methods to use instead of what I am using now? Especially relative to the matrix multiplication. Thank you in advance.

Reply:
> could you also explain me more practically what objects and methods to use instead of what I am using now?

Can you first demonstrate that there is a good reason for not just using Matrix here?

I realise that what you are doing involves sparse matrices but it also likely involves very large complicated expressions which tends to mean that the benefit of the sparse approach is lost in the weeds (a few extra zero multiplies makes no noticeable difference in context).

Reply:
> Could you point me where this method is implemented?

I linked it above already.